### ISSUE TITLE
65816 support

### ISSUE THREAD

**Issue by johnwbyrd** [llvm-mos/llvm-mos#32]
The 65816 shares a similar instruction architecture to its 8-bit counterpart. Register sizes are wider, and there are more addressing modes.

It seems likely that the 65816 wider register format, could speed up a number of llvm-mos operations.

I can think of several strategies for modifying or adding a back end specifically for the 816. We could lie to codegen and claim that the 65816 is actually a 32-bit processor, of which only the first 24 bits are used when rendering machine instructions.

**Comment by Molive-0** [comment]
Is there any easy way to implement handling changing the width of the registers from 8 to 16 bits and back? That seems like a niche feature.

**Comment by Selicre** [comment]
https://github.com/p4plus2/armadillo-spinach-muffin/ is a transpiler from i386 to 65816, which has a few tricks might be of interest to whoever wants to work on this.

**Comment by johnwbyrd** [comment]
> Is there any easy way to implement handling changing the width of the registers from 8 to 16 bits and back? That seems like a niche feature.

Not a quick or obvious one.  Codegen currently assumes that a, x, and y registers are 8 bits.  One way to overcome that limitation is to model larger 16-bit registers, of which a, x and y are subregisters.  This is possible, but not trivially so.

Re the transpiler: thanks for the link.  LLVM's GlobalISel already has really strong opinions about how to optimize for arbitrary architectures.  Strangely, performance optimization is not a limiting factor for the project at the moment -- generating consistently correct code is.

**Comment by johnwbyrd** [comment]
Memorializing a few thoughts on a 65816 version.

The address space for the MOS backend would have to detect that it's targeting a 65816, and so the address space for all instructions would have to change from 16 to 32 bits.  LLVM does not support address sizes that are not a power of two.  This in itself is not a trivial change.  As a matter of fact, it might be wise, first of all, to make sure that the 8-bit code generator would still continue to make the right choices, even if the address space for the compiler was 32 bits.

Once the current compiler survived all the smoke tests with the additional address space, it would then be time to add the additional 65816 instructions into the assembler.  At this point you could verify the 65816 opcodes, and make sure that assembly compiles and decompiles correctly.

After that, you could take a look at codegen itself.  Codegen would be the biggest issue,  The entire legalizer would have a radically different set of lowering rules from the 6502.  Many of the "lower everything to 8 bit" rules would be gone.  And there'd be a lot of "if (65816)" rules added here and there in codegen.

After codegen more or less works, then you would want to look at ways of intelligently segmenting 65816 code to make use of 16-bit addressing when possible, and ignoring the segment byte.  GlobalISel can help some with this, but ultimately we may ask for some help from the user to make sure the right code goes into the right code segments -- e.g. a default linker script will be provided, but performance freaks would want to write their own.

In short, it's not a weekend project.  But it could be done at some point, with a patient and organized effort.

**Comment by chorman0773** [comment]
For w65/65816, I'm working on my own toolchain (including eventually a C compiler, probably gcc). The support is being included in a fork of GNU binutils at https://github.com/chorman0773/binutils-gdb.
 I think it may be a good idea to, before support in either or both is too significantly locked in, to discuss the question of abi, to promote compatibility between the toolchains.
Some details of note:
- The target name used by config.sub for the 65816 processor is w65, with a default vendor of wdc. In patches submitted to config-patches@lists.gnu.org, I have sought to add 65816, w65c816, and wdc65c816 as aliases to w65. In the same patch series I also sought to add m6502 and m65c02 for NMOS 6502 and CMOS 6502.
- Support for the w65 elf target is using EM_WC65C816, 257. 
- I have a working document for an abi that can be used with the w65 processor, specified here: https://docs.google.com/document/d/1MYAY9Jn7BxH_FsQrGfZC_e5MjG9oCwgM2aYq4q9Nqrs/edit. Beyond the Section pertaining to Elf Files, much of this is open to comment and change. As Elf file support is already implemented in the previously linked fork of binutils, I do not recommend significant changes, beyond those relating to relaxation.

**Comment by johnwbyrd** [comment]
Instead of running off and creating yet another ELF convention for 6502, please note the following. https://github.com/johnwbyrd/llvm-mos/wiki/ELF-format-for-MOS-compatible-processors

**Comment by chorman0773** [comment]
Note that EM_WC65C816 (257) is a registered with generic-abi, (See the conversation here: https://groups.google.com/g/generic-abi/c/qaPzp2lRzDA, also note inclusion in upstream GNU Binutils: https://github.com/bminor/binutils-gdb/blob/14a6b9b4b68c33f1182c8e6060dcb25514268af9/include/elf/common.h#L354). While changing the relocations to match those of this project may be doable, my preference is to use the registration that already exists, rather than a value that, as far as I can tell, is not officially defined in any way.

**Comment by rdrpenguin04** [comment]
It is also worth noting that @chorman0773 has been developing his ABI for the past few years, though not very publicly. Connor, do you have that link?

There were various considerations made in this ABI to solve some of the same problems you have been having, and it would be interesting to compare ideas to see if some of them can be reconciled.

**Comment by chorman0773** [comment]
> Connor, do you have that link?

If you are referring to the ABI Working Document, I linked it above.

**Comment by rdrpenguin04** [comment]
Oh right, I'm blind.

**Comment by chorman0773** [comment]
Returning the discussion of 65816 vs. 6502 elf machine number from #86 (which primarily references the fact that EM_MCS6502 already existed and was registered with the maintainers of the ELF specification).
I would consider that 65816 vs. 6502 are different enough that they need different toolchain-level and programmer handling, similar to the EM_386 and EM_X86_64 distinction (albeit, those are different in two other respects, 386 uses ELFCLASS32 like m6502 and w65, whereas x86_64 uses ELFCLASS64). w65 needs a different set of relocations, and relaxations, from m6502. For example, the elf spec I wrote for it defines R_65C816_RELAX_JML, which can turn a jml instruction into a jmp or bra instruction. The m6502 wouldn't have such a link-relaxation because it doesn't have the jml instruction. Likewise, a format for m6502 wouldn't have use for R_65C816_REL16, R_65C816_ABS24, or R_65C816_BANK. As a more substantial distinction; R_MOS_ABS8 and R_65C816_DIRECT have different behaviour, R_MOS_ABS8 overflows with a value > 255, whereas R_65C816_DIRECT only overflows for a value > 65535 (and then only uses the low order byte).

**Comment by johnwbyrd** [comment]
Where is this ELF spec published?  Who has reviewed it so far?

**Comment by johnwbyrd** [comment]
The document you referenced above, literally refers to itself as an "SNES Dev ABI."  That alone should be disqualifying for its use as a generic standard across all 65xxx devices.  LLVM-MOS is merely a compiler, and it should not, by definition, have opinions as to which 65xxx hardware implementation it should run on.

As far as I can tell, your core argument here is that certain kinds of relocations only have meaning on certain kinds of hardware.  But it's up to the tools in all cases to figure out which features/bits should permit linkability between platforms.

Your R_65C816_RELAX_JML relocation is not a relocation.  It's a hint for some types of assemblers to make a relaxation.  LLVM handles instruction relaxation far earlier than the link step.

Do any C compilers or assemblers exist which follow your specification?

Have you reviewed our specification and relocation types yet?  We've already put together a list of relocations that should cover all LLVM's use cases, across the 65xx and the 658xx series of processors and clones.  See https://llvm-mos.org/wiki/ELF_specification for details.

**Comment by rdrpenguin04** [comment]
I will clarify a couple of things before chorman shows up to clarify the rest: the SNES-Dev ABI isn't meant for all 65xxx devices; it is meant for handling the 65816 processor on the SNES machine, similar to `x86_64-pc-linux-gnu` is meant for handling the amd64 processor, on a PC-compatible machine, running Linux, using the GNU ABI for all of those things.

The compiler does need to target the hardware; otherwise, how do you know where RAM and ROM are? How do you know what math-related hardware registers exist (for prior art, see x87 vs MMX vs SSE and variants)? Moreover, for a hosted system, how do you know how to output data to a screen? All of these, to some extent or another, depend on the machine.

While a PSABI should only target the processor, and there should be other ABI supplements for machine, the two are fairly intertwined in the case of the SNES, as well as (presumably) the IIGS. This does not preclude breaking the SNES-Dev ABI into a PSABI and a machine-specific supplement, it's just not what he chose; the three of us can work together to do that if it makes sense.

We do have a binutils fork that follows the specification; besides that, official GNU binutils and the `file` command both recognize `EM_65816`. A GCC fork is also in progress.

Let us know if you have any other questions!

**Comment by chorman0773** [comment]
Note that the above document is obsolete, and changed locations roughly a month ago. It is now in the SNES-Dev Project's github organization and repository, https://github.com/SNES-Dev/SNES-Dev/blob/main/docs/abi/v1.md. 
It has also been redesignated as a general ABI for w65 in general (rather than a SNES-specific Machine Supplicant) and in specific (rather than in general for all 65xx and 65xxx targets), though it remains a part of the SNES-Dev project as that is where it originated. 

I have indeed reviewed the ELF specification, and find it lacks a few operations, including link-time relaxations, which are notably not specific to the proposed w65 psABI (RISC-V's psABI also defines operations that can be relaxed at link time when the assembler or compiler may not have the necessary knowledge). The other missing operation's are R_WC65816_DIR, which is the same as (R_WC65816_ABS16)&0xff00, which is necessary for efficiently setting up the direct page for several adjacent accesses within the direct page, and R_WC65816_REL16, which is necessary to relocate the brl (branch long) instruction (which itself would be useful for position independent code). 

Link time relaxations are useful, especially for the w65 architecture where it can make sometimes meaningful differences in cycle counts and function lsizes, notably allowing relaxations on undefined symbols. 
Whether or not llvm may perform these operations is largerly irrelevent as the relaxations are optional - a valid implementation is to ignore them entirely - though this may leave llvm inferior to toolchains that do use them.

**Comment by johnwbyrd** [comment]
> Link time relaxations are useful, especially for the w65 architecture where it can make sometimes meaningful differences in cycle counts and function sizes, notably allowing relaxations on undefined symbols.

They may well be so, but LLVM on the whole does not yet understand them.  They are a uniquely compiler-specific gcc phenomenon as of this writing.  LLVM lacks the infrastructure to regenerate machine code in a (non-LTCG) link step, or to change the locations of machine code once it has been placed relative to the beginning of a section.  Therefore, no LLVM implementation can support these concepts, at least until the core of LLVM supports it.

65816 support is purely theoretical at this time for llvm-mos.  So is SWEET 16.  65816 codegen is not a weekend project, and I want to leave as much flexibility as possible to Daniel and others to do codegen for 65816 as they see fit in the future, if ever.  Our focus as of this writing, is getting 8-bit codegen clean, and leaving the door open to support other variants in the future.

Please help us on improving some of the more immediate 8-bit problems on this project, which will (in the fullness of time) improve any 24-bit support that we contemplate in the future.

**Comment by rdrpenguin04** [comment]
I believe chorman and I are trying to get something more-or-less nailed down so that, when we both do support codegen for 65c816, we don't have two mutually-incompatible ABIs. If you are not planning to work on this for a while, could you at least look at our ELF specification and give feedback on it, and then consider using it yourself later? That's our whole purpose of joining this discussion.

We may work on 8-bit codegen at some point, but that is not where our project is oriented at this time; as such, when/if we do get around to it, we would be trying to follow your existing ABI.

Let us know if you have any feedback; we are more than happy to help with ensuring future compatibility!

**Comment by rdrpenguin04** [comment]
"We both" referring to your group and the SNES-Dev group; I just realized that wasn't very clear.

**Comment by johnwbyrd** [comment]
> I believe chorman and I are trying to get something more-or-less nailed down so that, when we both do support codegen for 65c816, we don't have two mutually-incompatible ABIs. If you are not planning to work on this for a while, could you at least look at our ELF specification and give feedback on it, and then consider using it yourself later? That's our whole purpose of joining this discussion.

Do you have a functioning C compiler/assembler/linker, with source code, that we can look at?

**Comment by rdrpenguin04** [comment]
The repos are at the <https://github.com/SNES-Dev> organization; the mostly-functioning assembler is at [https://github.com/SNES-Dev/binutils-gdb](https://github.com/SNES-Dev/binutils-gdb)

**Comment by mysterymath** [comment]
> We may work on 8-bit codegen at some point, but that is not where our project is oriented at this time; as such, when/if we do get around to it, we would be trying to follow your existing ABI.

There may be a conflict in development methodologies between the two projects. I can only directly speak to the C side of this, but we don't provide any ABI stability or backwards-compatibility guarantees, and we're unlikely to for the considerable future. Our ABI isn't compatible with our ABI from 3 months ago, and it's reasonably likely that it won't be compatible with our ABI in 3 months' time.

Stabilizing an ABI without at least one high-quality implementation is dangerous, since the ABI might be constructed in such a way that non-obviously precludes high-quality implementations. History is littered with examples of this. Similarly, stabilizing an ABI with only one high-quality implementation is less dangerous, but it usually guarantees that there's only ever going to be one high-quality implementation of that thing.

The 65816 ABI we should use for llvm-mos should be one that guarantees that llvm-mos can be a high-quality compiler for that platform. I'll believe a given ABI satisfies this just as soon as someone shows me a high-quality compiler for it.

**Comment by rdrpenguin04** [comment]
That is definitely a fair argument, and that's why I'm suggesting we work together to make an ABI that works well for both of us. Connor's ABI is a great starting point, but it could probably use some growth and practical implementation.

For one, any ABI for any 65xx platform is going to be limited by trying to be compatible with itself; many cycles are wasted saving registers by calling convention and by forcing long jumps. Also, the SNES-Dev w65 ABI currently makes pointers 32-bit for the sake of alignment, and we're not sure if there's a way to make it 24-bit (or even, ideally, 16-bit and 24-bit) without causing some additional problem.

We'd love to hear input on the problem, and we want to collaborate; we don't want your implementation to lock into something bad any more than we do, and we don't want to lock out of each others' ABI any more either.

Connor also wanted to note that the ABI is not stabilized, and the intent is in fact to finish a compiler (likely GCC) before stabilizing the non-ELF-related sections.

The only other thing I have to say is that "Instead of running off and creating yet another ELF convention for 6502" sounds very hostile, even if it wasn't intended as such. We are not trying to stomp on your project, we are just trying to help make the best, compatible ABI we can.

Let us know if any of you have any questions!

**Comment by johnwbyrd** [comment]
We're not against working together.  Our issue is that, unless I am missing something, you do not seem to be ready to work with us yet.

Your order of operations to date, seems to have been:

1. Allocate some constants with Xinuos;
2. Add constants to binutils;
3. Write an "official" technical specification.

We believe strongly in a "code first" development strategy, and that strategy has paid off well for us. 

Our recommended order of operations is:

1. Get an assembler working;
2. Get ELF support sort of working;
3. Publish a working draft of a standard in the 6502 forums;
4. Get feedback on it;
5. Get the C compiler working (this should take you a year minimum);
6. Optimize the C compiler, without fixing the ABI;
7. Get an SDK working and stable;
8. Realize that a need exists to unify ELF support between LLVM-MOS and some other compiler (this is a big one -- literally no one is asking for this feature)
9. Figure out ways to unify constants while disturbing the smallest number of people;
10. Freeze the ABI (and we're not confident that we will ever do this);
11. Allocate some constants with Xinuos;
12. Unify the constants across compiler projects;
13. Write an official technical specification.

Please understand, we're not at all saying that you cannot be part of this project.  We welcome your reasoned code improvements and other contributions.  But it's absolutely the wrong time to begin unifying the toolchains' output formats.

The road to porting a compiler to the 65xx series, is littered with good intentions.  Do not underestimate how fantastically hard it is, to get a functional tool suite together.

We should revisit this issue once you have gotten your C compiler passing the gcc torture test suite.

**Comment by rdrpenguin04** [comment]
To be fair, we do have ELF support working, as well as a mostly working assembler in binutils. We choose to work to fit a specification so that we don't end up rewriting our code every time there's a significant problem; we can define first and find the problems there. Getting GCC working shouldn't be that difficult, it just hasn't been a priority for us; we've been using the assembler for a homebrew project, so the priority has been getting features into the assembler.

The constants were established for the specific purpose of beginning unification early; if the two of us design contrary systems, it will be hard to unify in the future. We're mainly hoping to solidify the ELF specification now, and we can revisit the C ABI later once one of us is closer to done with that.

Let us know if you have any other questions for us!

**Comment by chorman0773** [comment]
That is fair. We will get the gcc fbackend working and return then.

**Comment by rdrpenguin04** [comment]
Also, Connor mentioned to me that he will post his spec attempt on the 6502 forums for some feedback.

**Comment by mysterymath** [comment]
This SGTM; feel free to file issues for the uncompleted portion of this; this can be kept as a tracking bug for the work overall.

**Comment by asiekierka** [comment]
Done.

**Comment by Felice-Enellen** [comment]
Hey folks, former industry SNES gamedev here, and I just wanted to chime in on something said very early in this thread, which is that switching 8/16-bit register sizes is "a niche thing."

By now you may already know, but to be clear: it is often essential to good code, especially in terms of optimizing, partially because it effectively gives you a second "B" accumulator that you can switch to and from with the XAB (eXchange A, B) instruction. You absolutely will need good support for dynamic size switching in both code generation and optimization and should be making the heaviest use of it you can.

I'd give examples, but I haven't worked on 65816 since the 90s and my memory is pretty swiss-cheese about it at this point, sorry. Good luck, it's a worthy project. :)

**Comment by asiekierka** [comment]
We've had many discussions about this since, and I've done my own research - it is absolutely not niche for `A`, but it does seem to be rather niche for `X`/`Y`; which is good, as the latter are much, *much* harder to model in something generic like LLVM.

**Comment by johnwbyrd** [comment]
@asiekierka writes on Discord:

Either way, the 65C816 is hard but it's not, IMO, a fundamental rethinking of code generation, at least for a first pass: the most invasive changes are roughly divisible into three categories
(a) Adding support for the 24-bit address space; tracking data accesses and distinguishing between "near" and "far" jumps.
(b) Adding support for 16-bit registers; following in the footsteps of almost every other C compiler and most real-world 65C816 development, that means X/Y become permanently 16-bit, while A is switched at runtime.
(c) Adding support for the 65C816's hard stack; it has a 16-bit pointer and addressing modes that allow reading stack-offsetted variables, making this much more viable than on the 6502.
The rest is fairly minor, and some of it has already been done (like supporting the TXY/TYX opcodes).

**Comment by johnwbyrd** [comment]
Another piece of the puzzle has fallen into place: https://llvm-mos.org/wiki/DWARF_specification .  What this means is that, regardless of whatever convoluted calling convention we come up with, we now have an infrastructure to describe it to debuggers.

**Comment by Herringway** [comment]
I've been working on reverse engineering an SNES game using an obscure commercial compiler named VUCC. Perhaps the notes I've made on it may be of some use?

https://github.com/Herringway/ebsrc/blob/main/VUCC%20notes.md

**Comment by johnwbyrd** [comment]
Here is a discussion I had with Claude on this point.  Do not take it very seriously.  I am not even sure I agree with myself at this point.

----

# 65816 Support for LLVM-MOS: Architectural Plan

## Executive Summary

This document describes the architectural opinions for adding full 65816 support to the llvm-mos compiler. The guiding principle is simplicity for the programmer: **the 65816 should feel like a flat 16MB, 24-bit computer**.

### Key Decisions

| Decision | Choice | Rationale |
|----------|--------|-----------|
| Memory model | Flat 24-bit | Programmers think in flat memory, not banks |
| Pointer size | Always 24-bit (3 bytes) | No near/far confusion |
| Function calls | All JSL/RTL (far) | Correct by default |
| Register mode | Static 16-bit | Matches real-world usage |
| Optimization | Linker relaxation | Narrow addressing when provably safe |
| Address spaces | Not needed | Flat model eliminates the need |

### The Programmer's View

```c
// Just write normal C
void *p = malloc(100);      // 24-bit pointer, works everywhere
memcpy(dest, src, n);       // One memcpy, not near/far variants
callback();                 // Just works, even cross-bank

// sizeof(void*) == 3 is weird, but everything compiles
// No __near, no __far, no memory models to choose
// No "oops you used the wrong pointer type"
```

---

## The Core Challenges

### Challenge 1: Variable-Width Registers

The 65816's A/X/Y registers can be 8-bit or 16-bit, controlled by runtime processor flags:
- **m flag**: A register width (m=1: 8-bit, m=0: 16-bit)
- **x flag**: X/Y register width (x=1: 8-bit, x=0: 16-bit)

These flags change at runtime via SEP/REP instructions, and instruction encoding varies based on current mode.

**Decision: Static 16-bit mode compilation**

- Compiler flag selects mode: `-mcpu=w65816` implies 16-bit A/X/Y
- Entire compilation unit uses one mode
- SEP/REP only emitted at entry points or by explicit intrinsics
- This matches how most 65816 code is written in practice

### Challenge 2: 24-bit Address Space

The 65816 addresses 16MB via bank+offset (8+16 bits):
- 256 banks of 64KB each
- Bank 0 is special: direct page, hardware stack, interrupt vectors
- Program Bank Register (PBR/K): where code executes
- Data Bank Register (DBR/B): default bank for data access

**Decision: Flat 24-bit memory model**

All pointers are 24-bit. No near/far distinction. The compiler and linker handle the complexity.

---

## Memory Model: Flat 24-bit

### Data Layout

```llvm
; Data layout for 65816
target datalayout = "e-m:e-p:24:8-i16:8-i32:8-i64:8-..."
```

- `sizeof(void*)` = 3 (always)
- All memory accesses use long addressing by default
- All function calls use JSL/RTL by default
- Pointer arithmetic works across bank boundaries

### The Cost (Accepted for Correctness)

- Every pointer is 3 bytes (vs 2 on 6502)
- Every memory access instruction is 4 bytes (vs 3 for 16-bit addressing)
- Every call pushes 3-byte return address (vs 2)

### The Benefit

- Simple mental model for programmers
- No near/far pointer confusion
- Standard library just works (one `memcpy`, not two)
- Correct by default

### Why Not Near/Far Pointers?

We explicitly reject the traditional near/far pointer model:

1. **Cognitive burden**: Programmers must think about pointer types constantly
2. **Library duplication**: Need `memcpy` and `memcpy_far`, etc.
3. **Type system pollution**: `void*` vs `void __far*` throughout codebase
4. **Bug magnet**: Wrong pointer type = silent corruption or crash

The x86 world abandoned segmented memory for flat 32/64-bit and never looked back. We're applying the same philosophy.

### Why Not LLVM Address Spaces?

Address spaces (`addrspace(N)`) are useful when pointer types are *architecturally distinct* (like AVR's flash vs RAM). But on 65816:
- A 24-bit pointer can address anything
- Near vs far is an *optimization*, not a semantic distinction
- The flat model eliminates the need for type-level distinctions

If we later add `__near` as an expert escape hatch, it can be implemented without full address space machinery.

---

## Optimization Strategy: Linker Relaxation

### Phase 1: Make It Work

Emit conservative 24-bit code:
- All loads/stores use long addressing (`LDA $BBHHLL`)
- All calls use JSL/RTL
- All pointers are 3 bytes

This is slow but correct. A working compiler is the first milestone.

### Phase 2: Make It Fast

Linker relaxation narrows addressing when provably safe:

```
Before relaxation:
    LDA.l $7E1234    ; 4 bytes, long addressing
    JSL   some_func  ; 4 bytes, far call

After relaxation (if same bank):
    LDA.w $1234      ; 3 bytes, absolute addressing
    JSR   some_func  ; 3 bytes, near call
```

The linker has final symbol addresses and can prove:
- Is this data access in the same bank as DBR?
- Is this function call within the same bank as the caller?

### What About Struct Layout Optimization?

We considered narrowing pointers in structs (24-bit → 16-bit when provably safe), but rejected it:

1. **Struct layout is a type decision** - affects ABI
2. **Requires whole-program analysis** - can't prove pointer destinations in general
3. **Fragile** - one escaped pointer breaks the optimization
4. **Diminishing returns** - most bloat is in instructions, not struct pointers

**Decision**: Struct pointers are always 24-bit. Optimization focuses on instruction encoding.

### Comparison to PGO

This is similar to Profile-Guided Optimization:
1. **First pass**: Compile with worst-case assumptions
2. **Collect data**: Symbol table with final addresses
3. **Optimization**: Narrow based on known distances

But unlike PGO, the "profile" is deterministic (symbol addresses), not statistical. The optimization is guaranteed correct.

---

## Register Architecture

### 65816 Registers

| Register | Size | Notes |
|----------|------|-------|
| A (Accumulator) | 8 or 16-bit | Controlled by m flag |
| B (Hidden) | 8-bit | High byte when A is 16-bit (together: C) |
| X | 8 or 16-bit | Controlled by x flag |
| Y | 8 or 16-bit | Controlled by x flag |
| S (Stack) | 16-bit | Hardware stack pointer, bank 0 only |
| D (Direct Page) | 16-bit | Direct page base address |
| DBR | 8-bit | Data Bank Register |
| PBR | 8-bit | Program Bank Register |
| P (Status) | 8-bit | Flags including m and x |

### Register Definition Strategy

**Decision: Sub-register hierarchy**

```tablegen
def A : MOSReg<0, "A">;     // 8-bit (physical)
def B : MOSReg<1, "B">;     // High byte (hidden accumulator B)
def C : MOSRegWithSubRegs<0, "C", [A, B]>;  // 16-bit composite
```

This matches hardware reality:
- The 65816's "C" accumulator = A + hidden B byte
- XBA instruction swaps A and B
- Register allocator can reason about aliasing

### Imaginary Registers

The 6502 backend uses "imaginary" registers in zero page:
- **RC0-RC255**: 8-bit registers
- **RS0-RS127**: 16-bit pointer pairs

For 65816, add:
- **RL0-RL85**: 24-bit "long pointer" registers (3 bytes each)

The 65816's direct page can be relocated via the D register, giving more space than the 6502's fixed 256-byte zero page.

---

## Calling Convention

### Current 6502 ABI

- 8-bit values in A, X, then imaginary registers
- Pointers in RS1-RS7
- Soft stack for overflow

### Proposed 65816 ABI

**All calls use JSL/RTL (far calls):**
- 3-byte return address on hardware stack
- Function pointer size: 3 bytes
- Consistent everywhere

**Register usage in 16-bit mode:**
- 16-bit values in A (C), X, Y natively
- 24-bit pointers in RL registers
- Soft stack for locals (same as 6502)

**Linker can optimize to JSR/RTS when:**
- Caller and callee provably in same bank
- Saves 1 byte per call and 1 cycle

### DWARF CFI Implications

- Return address is always 3 bytes on hardware stack
- Simpler than mixed near/far (only one convention)
- CFI emission already supports this from 6502 debug work

---

## Platform Considerations: SNES

The SNES is a primary target. Does the flat 24-bit model work?

### SNES Memory Map

- Banks $00-$3F: LoROM area (interleaved ROM/RAM/hardware)
- Banks $7E-$7F: Work RAM (128KB)
- Banks $80-$FF: HiROM area
- Bank $00: Direct page, hardware stack, vectors

### How Flat 24-bit Interacts

**It works fine:**

1. **Code**: Can run from any bank. JSL/RTL handles cross-bank calls.
2. **Data**: Lives in WRAM ($7E-$7F). 24-bit pointers address it correctly.
3. **ROM**: Const data in ROM banks. 24-bit pointers work.
4. **Hardware stack**: Bank 0 only - used for return addresses (fine).
5. **Soft stack**: Anywhere (e.g., WRAM $7E). Platform SDK controls placement.

**Linker script example:**
```
.text   -> ROM (bank $00-$3F or $80-$FF)
.rodata -> ROM
.data   -> WRAM ($7E)
.bss    -> WRAM ($7E)
```

### Code Size Concerns

SNES cartridges have limited ROM. The flat model's 4-byte instructions (vs 3-byte) matter.

**Mitigation**: Linker relaxation recovers most of the savings. The linker knows final addresses and can narrow to 3-byte absolute addressing when data is in the same bank as DBR.

---

## Linker Constraints

### Functions Cannot Cross Bank Boundaries

When PC increments past $xxFFFF, it wraps to $xx0000 in the same bank - it doesn't increment the bank byte.

**Linker rule**: No function can span a bank boundary.

- Functions are placed entirely within one bank
- Error if a function exceeds 64KB (pathological case)
- Padding/alignment at bank boundaries if needed

This is similar to page alignment on other architectures. For normal code, it's invisible.

### Branch Range Limitations

- BRA/Bcc: ±128 bytes
- BRL: ±32KB

The assembler/linker already handles out-of-range branches by converting to jump sequences. Existing infrastructure.

---

## Implementation Approach

### Same Backend, Different Subtarget

65816 is a **subtarget** within the MOS backend (like ARM vs Thumb):

```
Triple: mos-unknown-unknown (same)
CPU: -mcpu=mos6502 (default) vs -mcpu=w65816
```

The existing `FeatureW65816` is already defined. Most code paths check `Subtarget->hasW65816()`.

### Code Reuse

| Component | Reuse | Notes |
|-----------|-------|-------|
| Instruction encoding | Yes | 65816 is superset of 6502 |
| Assembler/disassembler | Yes | Already supports 65816 mnemonics |
| ELF object format | Yes | Same format |
| Linker (lld) | Yes | Relaxation framework exists |
| Imaginary registers | Partial | Add RL class for 24-bit |
| Soft stack | Yes | Same architecture |
| GlobalISel | Partial | Need 16-bit legalization |

### ABI Incompatibility

Different pointer sizes (16-bit vs 24-bit) make 6502/65816 ABI interop impractical:
- Different `sizeof(void*)`
- Different struct layouts
- Different calling conventions

**Decision**: 65816 is a clean break. Separate libc/SDK builds. ELF markers prevent accidental mixing.

---

## Implementation Phases

### Phase 0: Architecture Design (Current)

- [x] Decide on compilation model → Static 16-bit mode
- [x] Decide on memory model → Flat 24-bit
- [x] Decide on optimization strategy → Linker relaxation
- [ ] Document ABI for 65816 mode
- [ ] Define DWARF register numbers for new registers (B, D, DBR, PBR)

### Phase 1: Foundation (Make It Work)

**Goal**: Compiler produces working 65816 code, even if slow

- [ ] Add `-mcpu=w65816` target with 24-bit data layout
- [ ] Define 16-bit register variants (C=A+B, X16, Y16)
- [ ] Add RL (24-bit long pointer) imaginary register class
- [ ] All loads/stores use long addressing
- [ ] All calls use JSL/RTL

### Phase 2: Type System

- [ ] Modify legalizer to support native i16 for A/X/Y
- [ ] Handle i24 pointers (or i32 with truncation)
- [ ] Update register bank info for 16-bit and 24-bit values

### Phase 3: Instruction Support

- [ ] Enable existing 65816 instructions (JSL, RTL, BRL, etc.)
- [ ] Add 16-bit variants of ALU instructions
- [ ] Long addressing modes for all memory operations
- [ ] Bank register instructions (PHB/PLB, PHK, etc.)

### Phase 4: Calling Convention

- [ ] Define `mos65816cc` calling convention
- [ ] Update prologue/epilogue for 3-byte returns
- [ ] 24-bit function pointer indirect calls

### Phase 5: Optimization (Make It Fast)

- [ ] Linker relaxation for addressing modes (LDA.l → LDA.w)
- [ ] Linker relaxation for calls (JSL → JSR when same bank)
- [ ] MVN/MVP block move optimization
- [ ] XBA optimization (swap A/B bytes)

---

## What Already Exists

The llvm-mos codebase already has 65816 groundwork:

- `FeatureW65816` subtarget feature
- `FamilyW65816` device family
- 65816-specific instructions defined (JSL, RTL, BRL, MVN, MVP, PEA, etc.)
- Long addressing modes defined (AbsoluteLong, IndirectLong)
- TSFlags for M/X mode bits (MLow, MHigh, XLow, XHigh)
- `hasW65816()` and `hasW65816Or65EL02()` predicates

### What's Missing

- 16-bit register definitions (C, X16, Y16)
- Bank registers (DBR, PBR) in register file
- Direct page register (D) in register file
- Native 16-bit type legalization
- 16-bit instruction selection patterns
- Mode-aware calling convention
- DWARF register numbers for new registers
- 24-bit pointer data layout

---

## Risk Assessment

| Decision | Risk | Mitigation |
|----------|------|------------|
| Static 16-bit mode | Users want dynamic switching | Document limitations; add per-function attributes later |
| Flat 24-bit model | Slower than optimized near/far | Linker relaxation recovers most performance |
| 24-bit imaginary registers (RL) | Direct page pressure | 65816 has more DP space than 6502 ZP |
| Sub-register hierarchy | Register allocator complexity | Follow ARM/x86 patterns; extensive testing |
| i24 type handling | LLVM doesn't natively support i24 | Use i32 with truncation initially |
| All JSL/RTL calls | Overhead vs JSR/RTS | Linker relaxation to JSR when same bank |

---

## Key Files to Modify

| File | Changes |
|------|---------|
| `llvm/lib/Target/MOS/MOSRegisterInfo.td` | Add C, B, X16, Y16, D, DBR, PBR, RL0-RL85 |
| `llvm/lib/Target/MOS/MOSInstrInfo.td` | Enable 65816 instructions, 16-bit patterns |
| `llvm/lib/Target/MOS/MOSCallingConv.td` | New `mos65816cc` convention |
| `llvm/lib/Target/MOS/MOSLegalizerInfo.cpp` | Native i16, i24 pointer legalization |
| `llvm/lib/Target/MOS/MOSInstructionSelector.cpp` | 16-bit register class selection |
| `llvm/lib/Target/MOS/MOSSubtarget.h/cpp` | 65816 subtarget, data layout selection |
| `llvm/lib/Target/MOS/MOSFrameLowering.cpp` | 3-byte return address handling |
| `llvm/lib/Target/MOS/MOSCallLowering.cpp` | JSL/RTL for far calls |
| `llvm/lib/Target/MOS/MOSDevices.td` | W65816 device with features |
| `clang/lib/Basic/Targets/MOS.cpp` | 24-bit pointer data layout |
| `lld/ELF/Arch/MOS.cpp` | Bank-aware relaxation |

---

## Open Questions

1. **i24 representation**: LLVM doesn't have native i24. Options:
   - Use i32 and truncate (wasteful but simple)
   - Custom i24 handling (complex)
   - Needs investigation of what other backends do

2. **Direct page relocation**: The D register can point anywhere in bank 0. How do we communicate the DP base to the linker for relaxation?

3. **65C816 vs W65C816**: Are there any behavioral differences that affect code generation?

---

## Summary

The 65816 support strategy is:

1. **Flat 24-bit memory model** - programmers never think about banks
2. **All pointers are 24-bit** - no near/far, sizeof(void*) == 3
3. **All calls use JSL/RTL** - correct by default
4. **Linker relaxation** - optimize instruction encoding post-link
5. **Same backend, different subtarget** - maximize code reuse
6. **Correct first, fast later** - working compiler is milestone 1

This approach prioritizes programmer ergonomics over micro-optimization. The x86 world proved that flat memory models win. We're applying that lesson to a weird old 8/16-bit CPU.

**Comment by gingerbeardman** [comment]
Sharing a small experiment that fits under this umbrella: native 16-bit accumulator codegen for the 65816, behind an off-by-default `+native-16bit-accumulator` feature. i16 `add/sub/and/or/xor` become `M=0` `REP/SEP` regions instead of byte-at-a-time. The C ABI, pointer model, 8-bit `X`/`Y` and banking are all unchanged, so it interops with stock objects.

Short version: on long i16 chains it's a clear win (~1.35x, 55% less code), but on real scattered i16 code it actually regresses unless it's gated by a profitability heuristic — which ended up being most of the work. I wrote that up properly, with the cautionary benchmark, over in #321.

Branch (pinned to `c798c314`):
<https://github.com/gingerbeardman/llvm-mos/tree/mos-native-16bit-accumulator>
Write-up + reproduce:
<https://gist.github.com/gingerbeardman/03d232f637c02a30b16c42a302cf782c>

This is only the accumulator (`M` flag) — 16-bit `X`/`Y` and the 24-bit address space (#320) are still open.
